receiver/nvmlreceiver/client.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !windows
+// +build !windows
+
+package nvmlreceiver
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"go.uber.org/zap"
+)
+
+const maxWarningsForFailedDeviceMetricQuery = 5
+
+type nvmlClient struct {
+	logger                         *zap.SugaredLogger
+	disable                        bool
+	handleCleanup                  func()
+	devices                        []nvml.Device
+	devicesModelName               []string
+	deviceToLastSeenTimestamp      map[nvml.Device]uint64
+	deviceMetricToFailedQueryCount map[string]uint64
+}
+
+type nvmlMetric struct {
+	time  time.Time
+	gpuID uint
+	name  string
+	value [8]byte
+}
+
+// calling nvml.Init() twice causes an unnecessary error (also wrap here for mocking)
```

Comment:
Hm, if I recall right, yes. Though logically it's conceivable someone might want more than one NVML clients within the same process. Granted, it may never happen, but it makes sense. So I think it's fair to put this here. And looks clean to me to do it this way. Why, did you have some ideas?